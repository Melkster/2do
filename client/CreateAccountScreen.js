import React, { Component } from "react";
import {
  Alert,
  Button,
  TouchableWithoutFeedback,
  Keyboard,
  KeyboardAvoidingView,
  Text,
  TextInput,
  Modal,
  View
} from "react-native";
import socket from "./socket";
import styles from "./styles";

export default class LogInScreen extends Component {
  constructor(props) {
    super(props);

    this.state = {
      username: "",
      password1: "",
      password2: "",
      modalVisible: false,
      credentialsStatus: ""
    };

    socket.on("register", (_, err) => this._handleRegister(err));
  }

  static navigationOptions = {
    title: "Create account"
  };

  // Remove socket listeners on unmount
  componentWillUnmount() {
    socket.removeAllListeners();
  }

  render() {
    return (
      <TouchableWithoutFeedback onPress={Keyboard.dismiss} accessible={false}>
        <KeyboardAvoidingView behavior="height" style={styles.container} onPress={Keyboard.dismiss}>
          <TextInput
            value={this.state.username}
            onChangeText={username => this._setStateAndUpdateStatus({ username })}
            placeholder={"Username"}
            style={styles.input}
          />
          <TextInput
            value={this.state.password1}
            onChangeText={password1 => this._setStateAndUpdateStatus({ password1 })}
            placeholder={"Password"}
            secureTextEntry={true}
            style={styles.input}
          />
          <TextInput
            value={this.state.password2}
            onChangeText={password2 => this._setStateAndUpdateStatus({ password2 })}
            placeholder={"Please repeat your password"}
            secureTextEntry={true}
            style={styles.input}
          />

          <Text style={styles.errorStatusIndicator}>{this._capitalize(this.state.credentialsStatus)}</Text>

          <Button title={"Create account"} style={styles.input} onPress={this._confirm} />

          <Modal visible={this.state.modalVisible} animationType={"slide"} onRequestClose={this._closeModal}>
            <View style={styles.container}>
              <Text>You successfully created an account</Text>
              <Button onPress={this._closeModal} title="Back to login" />
            </View>
          </Modal>
        </KeyboardAvoidingView>
      </TouchableWithoutFeedback>
    );
  }

  _setStateAndUpdateStatus = state => {
    this.setState(state, () => {
      return this._credentialsStatus();
    });
  };

  _validateCredentials = (username, password1, password2) => {};

  /**
   * Sets this.state.credentialsStatus.
   *
   * Checks
   */
  _credentialsStatus = status => {
    if (this.state.password1 && this.state.password2 && this.state.password1 !== this.state.password2) {
      if (this.state.password1 !== this.state.password2) {
        status = "Passwords don't match";
      }
    }

    if (status !== this.state.credentialsStatus) {
      // Check if this.state.credentialsStatus will change to prevent infinite recursive `setState()` calls
      this.setState({ credentialsStatus: status });
    }
  };

  /**
   * Sets the visibility of modal.
   */
  _setModalVisible = visible => {
    this.setState({ modalVisible: visible });
  };

  _confirm = () => {
    this._registerUser();
  };

  /**
   * Closes modal and navigates back to the initial screen. Also calls the
   * `setFiels()` callback function, which passes `username` and `password` to
   * the initial screen.
   */
  _closeModal = () => {
    this._setModalVisible(false);
    this.props.navigation.popToTop();
    this.props.navigation.state.params.setFields(this.state.username, this.state.password1);
  };

  /**
   * Sends a `register` event if `username` and `password` are defined.
   */
  _registerUser = callback => {
    if (this.state.password1 === this.state.password2) {
      socket.emit("register", this.state.username, this.state.password1);
    }
  };

  /**
   * Registers user.
   *
   * If there is no `err`, displays confirmation modal.
   */
  _handleRegister = err => {
    if (err) this._credentialsStatus(err);
    else this._setModalVisible(true);
  };

  /**
   *  Capitalizes the first letter in `string` if `string` is defined.
   */
  _capitalize = string => {
    return string ? string.charAt(0).toUpperCase() + string.substring(1) : string;
  };
}
