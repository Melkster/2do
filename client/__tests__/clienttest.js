import "react-native";
import React from "react";
import CreateAccountScreen from "../CreateAccountScreen";

import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer.create(<CreateAccountScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
