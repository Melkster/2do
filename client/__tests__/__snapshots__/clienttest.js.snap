// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={false}
  onLayout={[Function]}
  onPress={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "margin": 10,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Username"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "gray",
        "borderWidth": 1,
        "height": 44,
        "marginBottom": 10,
        "padding": 10,
        "width": 300,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Password"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Object {
        "borderColor": "gray",
        "borderWidth": 1,
        "height": 44,
        "marginBottom": 10,
        "padding": 10,
        "width": 300,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Please repeat your password"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Object {
        "borderColor": "gray",
        "borderWidth": 1,
        "height": 44,
        "marginBottom": 10,
        "padding": 10,
        "width": 300,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text
    style={
      Object {
        "color": "red",
        "marginBottom": 10,
      }
    }
  >
    
  </Text>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Create account
      </Text>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "margin": 10,
        }
      }
    >
      <Text>
        You successfully created an account
      </Text>
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Back to login
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
